iree/compiler/Dialect/Util/Transforms/GreedyHoistIntoGlobals.cpp

```diff
@@ -0,0 +1,308 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Util/Analysis/Constant/ConstExpr.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Dialect/Util/Transforms/Passes.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Analysis/SliceAnalysis.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+#include "mlir/Dialect/Linalg/IR/LinalgOps.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/SymbolTable.h"
+
+#define DEBUG_TYPE "iree-greedy-hoist-into-globals"
+
+using llvm::dbgs;
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Util {
+namespace {
+
+// Maps an original value in the program to the symbol name of a global.
+using HoistedValueMap = llvm::DenseMap<Value, GlobalOp>;
+
+bool canHoistOperand(OpOperand *operand) {
+  Operation *op = operand->getOwner();
+  // For linalg ops, we only want to hoist inputs.
+  if (auto structuredOp = dyn_cast<linalg::LinalgOp>(op)) {
+    return operand->getOperandNumber() < structuredOp.getNumInputs();
+  }
+
+  // Fallback to yes.
+  return true;
+}
+
+bool isHoistableLeaf(const ConstExprAnalysis::ConstValueInfo *info) {
+  // Generally, we prefer to not hoist broadcasts.
+  if (auto genericOp = dyn_cast<linalg::GenericOp>(info->getOperation())) {
+    // Detect op that only broadcast input as fusing them makes the new
+    // op cheaper.
+    if (genericOp.getNumParallelLoops() == genericOp.getNumLoops() &&
+        isa<linalg::YieldOp>(genericOp.getBody()->front())) {
+      for (OpOperand *opOperand : genericOp.getInputOperands()) {
+        AffineMap indexingMap = genericOp.getTiedIndexingMap(opOperand);
+        if (indexingMap.isProjectedPermutation() &&
+            indexingMap.getNumDims() != indexingMap.getNumResults()) {
+          return false;
+        }
+      }
+    }
+  }
+
+  return true;
+}
+
+// Greedily hoists all eligible constants that escape to non constant
+// expressions into globals. It is not expected that such a greedy algorithm
+// is great, but it is simple. Naive use of this algorithm very likely
+// favors programs that consume more memory at runtime than is strictly
+// necessary. Either this algorithm can be made smarter or a follow-on pass
+// can sink globals into the program where it is profitable to reduce
+// working set size.
+class GreedyHoistIntoGlobalsPass
+    : public PassWrapper<GreedyHoistIntoGlobalsPass,
+                         OperationPass<mlir::ModuleOp>> {
+ public:
+  StringRef getArgument() const override {
+    return "iree-greedy-hoist-into-globals";
```

Comment:
we've got a lot of passes so we namespace them to keep listings and such nicely sorted:
```suggestion
    return "iree-util-greedy-hoist-into-globals";
```